egui/src/widgets/plot/items.rs

```diff
@@ -1446,3 +1560,947 @@ impl PlotItem for PlotImage {
+}
+
+fn rulers_color(ui: &Ui) -> Color32 {
+    if ui.visuals().dark_mode {
+        Color32::from_gray(100).additive()
+    } else {
+        Color32::from_black_alpha(180)
+    }
+}
+
+fn vertical_line(pointer: Pos2, transform: &ScreenTransform, line_color: Color32) -> Shape {
+    let frame = transform.frame();
+    Shape::line_segment(
+        [
+            pos2(pointer.x, frame.top()),
+            pos2(pointer.x, frame.bottom()),
+        ],
+        (1.0, line_color),
+    )
+}
+
+fn horizontal_line(pointer: Pos2, transform: &ScreenTransform, line_color: Color32) -> Shape {
+    let frame = transform.frame();
+    Shape::line_segment(
+        [
+            pos2(frame.left(), pointer.y),
+            pos2(frame.right(), pointer.y),
+        ],
+        (1.0, line_color),
+    )
+}
+
+/// Draws a cross of horizontal and vertical ruler at the `pointer` position.
+/// `value` is used to for text displaying X/Y coordinates.
+#[allow(clippy::too_many_arguments)]
+pub(super) fn rulers_at_value(
+    pointer: Pos2,
+    value: Value,
+    name: &str,
+    plot: &PlotConfig<'_>,
+    shapes: &mut Vec<Shape>,
+) {
+    let line_color = rulers_color(plot.ui);
+    if plot.show_x {
+        shapes.push(vertical_line(pointer, plot.transform, line_color));
+    }
+    if plot.show_y {
+        shapes.push(horizontal_line(pointer, plot.transform, line_color));
+    }
+
+    let mut prefix = String::new();
+
+    if !name.is_empty() {
+        prefix = format!("{}\n", name);
+    }
+
+    let text = {
+        let scale = plot.transform.dvalue_dpos();
+        let x_decimals = ((-scale[0].abs().log10()).ceil().at_least(0.0) as usize).at_most(6);
+        let y_decimals = ((-scale[1].abs().log10()).ceil().at_least(0.0) as usize).at_most(6);
+        if plot.show_x && plot.show_y {
+            format!(
+                "{}x = {:.*}\ny = {:.*}",
+                prefix, x_decimals, value.x, y_decimals, value.y
+            )
+        } else if plot.show_x {
+            format!("{}x = {:.*}", prefix, x_decimals, value.x)
+        } else if plot.show_y {
+            format!("{}y = {:.*}", prefix, y_decimals, value.y)
+        } else {
+            unreachable!()
+        }
+    };
+
+    shapes.push(Shape::text(
+        plot.ui.fonts(),
+        pointer + vec2(3.0, -2.0),
+        Align2::LEFT_BOTTOM,
+        text,
+        TextStyle::Body,
+        plot.ui.visuals().text_color(),
+    ));
+}
+
+fn find_closest_rect<'a, T>(
+    rects: impl IntoIterator<Item = &'a T>,
+    point: Pos2,
+    transform: &ScreenTransform,
+) -> Option<HoverElement>
+where
+    T: 'a + RectElement,
+{
+    rects
+        .into_iter()
+        .enumerate()
+        .map(|(index, bar)| {
+            let bar_rect: Rect = transform.rect_from_values(&bar.bounds_min(), &bar.bounds_max());
+            let dist_sq = point.distance_from_rect_sq(bar_rect);
+
+            HoverElement { index, dist_sq }
+        })
+        .min_by_key(|e| e.dist_sq.ord())
+}
```

Comment:
There's a lot of new code here. I think it would make sense to add them to new files (maybe `egui/src/widgets/plot/items/bar.rs` and `egui/src/widgets/plot/items/box.rs`)